OSDOCS-11182: Ingress controller feature MicroShift (#84542)
openshift/openshift-docs: modules/microshift-ingress-controller-config.adoc

. Replace the default values in the `network` section of the {microshift-short} YAML with your valid values, or create a configuration snippet file with the sections you need.
+
.Ingress controller configuration fields with default values
[source,yaml]
----
apiServer:
# ...
ingress:
    defaultHTTPVersion: 1
    forwardedHeaderPolicy: Append
    httpCompression:
        mimeTypes:
            - ""
    httpEmptyRequestsPolicy: Respond
# ...
    logEmptyRequests: Log
# ...
    tuningOptions:
        clientFinTimeout: 1s
        clientTimeout: 30s
        headerBufferBytes: 0
        headerBufferMaxRewriteBytes: 0
        healthCheckInterval: 5s
        maxConnections: 0
        serverFinTimeout: 1s
        serverTimeout: 30s
        threadCount: 4
        tlsInspectDelay: 5s
        tunnelTimeout: 1h
# ...
----
+
.Ingress controller configuration fields definitions table
[cols="3a,8a",options="header"]
|===
|Parameter |Description

|`defaultHTTPVersion`
|Sets the HTTP version for the ingress controller. Default value is `1` for HTTP 1.1.

> //Q: which versions do we support; HTTP/1.0, 2 and 3? do we need to configure a load balancer for 2 and 3?

Reviewer: we support by default  http/1.1 and http/2
see also  https://docs.openshift.com/container-platform/4.17/networking/networking_operators/ingress-operator.html - `Enabling HTTP/2 Ingress connectivity`

+
[IMPORTANT]
====
Setting this field is not recommended because increasing the number of `HAProxy` threads allows ingress controller pods to use more CPU time under load, and prevent other pods from receiving the CPU resources they need to perform.
====

* The `tuningOptions.tlsInspectDelay` parameter specifies how long the router can hold data to find a matching route. Setting this value too short can cause the router to fall back to the default certificate for edge-terminated, re-encrypted, or passthrough routes, even when using a better-matched certificate. The default inspect delay is `5s`.

* The `tuningOptions.tunnelTimeout` parameter specifies how long a tunnel connection, including websockets, remains open while the tunnel is idle. The default timeout is `1h`.
|===

. Complete any other configurations you require, then start or restart {microshift-short} by running one the following commands:
+
[source,terminal]
----
$ sudo systemctl start microshift
----
+
[source,terminal]
----
$ sudo systemctl restart microshift
----

> //Q: Anything else we need to do? e.g., other settings we need to make on the network side?

Reviewer: only router pod restart  which should be done automatically , see below the verification.

> //Q: Any verification steps we should make before we restart MicroShift?

Reviewer: after changing ingress configuration  and perform service restart ,
the ingress pod should be restarted with the changes in place,
so we can check if the pod is restarted, using the command:
```bash
> oc get pods -n openshift-ingress
NAME                              READY   STATUS    RESTARTS   AGE
router-default-8649b5bf65-w29cn   1/1     Running   0          6m10s
```
one should look at the AGE field which should be recent

// Module included in the following assemblies:
//
// * microshift_configuring/microshift-ingress-controller.adoc

:_mod-docs-content-type: PROCEDURE
[id="microshift-ingress-control-config_{context}"]
= Configuring ingress control in {microshift-short}

You can use detailed ingress control settings by updating the {microshift-short} service configuration file.

.Prerequisites

* You installed the OpenShift CLI (`oc`).
* You have root access to the cluster.
* Your cluster uses the OVN-Kubernetes network plugin.
//Q: Any other prereqs?

.Procedure

. You can apply ingress control settings in either of the two following places:

.. Update the {microshift-short} `config.yaml` configuration file by making a copy of the provided `config.yaml.default` file in the `/etc/microshift/` directory.

.. Use a configuration snippet to apply the ingress control settings you want. To do this, create a configuration snippet YAML file and put it in the the `/etc/microshift/config.d/` configuration directory.

> 

Reviewer 2: There are some indentation issues with this note, causing the next step to restart the order as step 1. At the very least, I think there needs to be a + added right above the note: 
```suggestion
+
```

However, I think that would indent the note under step 1b, but it seems like the note applies to both substeps. I don't know how you'd want to handle that, but at least this way the indentation and number ordering is valid

[NOTE]
====
* After you create it, the `config.yaml` file takes precedence over built-in settings. The configuration file is read every time the {microshift-short} service starts.
* Configuration snippet YAMLs take precedence over all settings anda `config.yaml` configuration file. See the Additional resources links for more information.
====

> . When using the `config.yaml`, replace the default values in the `network` section of the {microshift-short} YAML with your valid values.

Reviewer 2: Suggestion - if this is meant to be a conditional step and only applies to users who chose step 1a, I think it would be a little clearer to use "if" instead of "when": 
```suggestion
. If using the `config.yaml`, replace the default values in the `network` section of the {microshift-short} YAML with your valid values.
```

But I don't know the context of this procedure so i'll leave that to you

Author: Yeah, revised that, thanks. There are only two ways to configure MicroShift itself; either by using the config YAML, or by using config snippets. The snippets are new for 4.18, so still working out how to best include them as an option.

+
.Ingress controller configuration field snippet with default values
[source,yaml]
----
apiServer:
# ...
ingress:
    defaultHTTPVersion: 1
    forwardedHeaderPolicy: Append
    httpCompression:
        mimeTypes:
            - ""
    httpEmptyRequestsPolicy: Respond
# ...
    logEmptyRequests: Log
# ...
    tuningOptions:
        clientFinTimeout: 1s
        clientTimeout: 30s
        headerBufferBytes: 0
        headerBufferMaxRewriteBytes: 0
        healthCheckInterval: 5s
        maxConnections: 0
        serverFinTimeout: 1s
        serverTimeout: 30s
        threadCount: 4
        tlsInspectDelay: 5s
        tunnelTimeout: 1h
# ...
----

> 

Reviewer 2: This needs a + so that the table is indented with the step and so that the last step doesn't restart as step 1:
```suggestion
+
```

[cols="3a,8a",options="header"]
|===
|Parameter |Description

|`defaultHTTPVersion`

> |`defaultHTTPVersion` sets the HTTP version for the Ingress Controller. Default value is `1`.

Reviewer 2: Just a suggestion to consider - for each description like this one that starts with the name of the parameter that's already mentioned in the adjacent cell, I feel like you could remove the name from the start of the description. This way, you'd avoid the slightly awkward formatting of starting a sentence with a lowercase letter because it's a parameter name:
```suggestion
|Sets the HTTP version for the Ingress Controller. Default value is `1`.
```

Up to you though

//Q: which versions do we support; HTTP/1.0, 2 and 3? do we need to configure a load balancer for 2 and 3?

|`forwardedHeaderPolicy`
|By setting the `forwardedHeaderPolicy`, you specify when and how the Ingress Controller sets the `Forwarded`, `X-Forwarded-For`, `X-Forwarded-Host`, `X-Forwarded-Port`, `X-Forwarded-Proto`, and `X-Forwarded-Proto-Version` HTTP headers. The following values are valid:

* `Append`, preserves any existing headers by specifying that the Ingress Controller appends them.
* `Replace`, removes any existing headers by specifying that the Ingress Controller sets the headers.
* `IfNone` sets the headers set by specifying that the Ingress Controller sets the headers  if they are not already set.
* `Never`, preserves any existing headers by specifying that the Ingress Controller never sets the headers.

|`httpCompression`
|`httpCompression` defines the policy for HTTP traffic compression.

> * `httpCompression.mimeTypes` defines a list of or a single MIME type to which compression is applied. For example, `text/css; charset=utf-8`, `text/html`, `text/*`, `image/svg+xml`, `application/octet-stream`, `X-custom/customsub`, using the format pattern, `type/subtype; [;attribute=value]`. The `types` are: application, image, message, multipart, text, video, or a custom type prefaced by `X-`. To see the full notation for MIME types and subtypes, see link:https://datatracker.ietf.org/doc/html/rfc1341#page-7[RFC1341]

Reviewer 2: Missing a period here: 
```suggestion
* `httpCompression.mimeTypes` defines a list of or a single MIME type to which compression is applied. For example, `text/css; charset=utf-8`, `text/html`, `text/*`, `image/svg+xml`, `application/octet-stream`, `X-custom/customsub`, using the format pattern, `type/subtype; [;attribute=value]`. The `types` are: application, image, message, multipart, text, video, or a custom type prefaced by `X-`. To see the full notation for MIME types and subtypes, see link:https://datatracker.ietf.org/doc/html/rfc1341#page-7[RFC1341].
```

> * `httpCompression.mimeTypes` defines a list of or a single MIME type to which compression is applied. For example, `text/css; charset=utf-8`, `text/html`, `text/*`, `image/svg+xml`, `application/octet-stream`, `X-custom/customsub`, using the format pattern, `type/subtype; [;attribute=value]`. The `types` are: application, image, message, multipart, text, video, or a custom type prefaced by `X-`. To see the full notation for MIME types and subtypes, see link:https://datatracker.ietf.org/doc/html/rfc1341#page-7[RFC1341]

Reviewer 2: Also, per the [docs guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#lead-in-link-sentence-style-for-the-link-attribute), when linking to external sites we should add a description of the location in parentheses following the link: 
```suggestion
* `httpCompression.mimeTypes` defines a list of or a single MIME type to which compression is applied. For example, `text/css; charset=utf-8`, `text/html`, `text/*`, `image/svg+xml`, `application/octet-stream`, `X-custom/customsub`, using the format pattern, `type/subtype; [;attribute=value]`. The `types` are: application, image, message, multipart, text, video, or a custom type prefaced by `X-`. To see the full notation for MIME types and subtypes, see link:https://datatracker.ietf.org/doc/html/rfc1341#page-7[RFC1341] (IETF Datatracker documentation).
```

I'm guessing the site description/name here, so please change if needed

|`httpEmptyRequestsPolicy`
|`httpEmptyRequestsPolicy` describes how HTTP connections are handled if the connection times out before a request is received. Allowed values for this field are `Respond` and `Ignore`. The default value is `Respond`. The following are valid values:

> * `Respond`, causes the ingress controller sends an HTTP `400` or `408` response, logs the connection if access logging is enabled, and counts the connection in the appropriate metrics.

Reviewer 2: ```suggestion
* `Respond`, causes the ingress controller to send an HTTP `400` or `408` response, logs the connection if access logging is enabled, and counts the connection in the appropriate metrics.
```

> * `Ignore`, adds the `http-ignore-probes` parameter in the `HAproxy configuration. If the field is set to `Ignore`, the ingress controller closes the connection without sending a response, then logs the connection or incrementing metrics.

Reviewer 2: I think this is missing a backtick and messing up the formatting in the rendered preview. Looks like it might belong here?
```suggestion
* `Ignore`, adds the `http-ignore-probes` parameter in the `HAproxy` configuration. If the field is set to `Ignore`, the ingress controller closes the connection without sending a response, then logs the connection or incrementing metrics.
```

The following are valid values:

* `Log`, indicates that an event should be logged.
* `Ignore`, sets the `dontlognull` option in the `HAproxy` configuration.

|`tuningOptions`
|`tuningOptions` specifies options for tuning the performance of ingress controller pods.

* `tuningOptions.clientFinTimeout` specifies how long a connection is held open while waiting for the client response to the server closing the connection. The default timeout is `1s`.

* `tuningOptions.clientTimeout` specifies how long a connection is held open while waiting for a client response. The default timeout is `30s`.

* `tuningOptions.headerBufferBytes` specifies how much memory is reserved, in bytes, for Ingress Controller connection sessions. This value must be at least `16384` if HTTP/2 is enabled for the Ingress Controller. If not set, the default value is `32768` bytes.
+
[IMPORTANT]
====

> Setting this field not recommended because `headerBufferMaxRewriteBytes` values that are too small can break the ingress controller. Conversely, values for `headerBufferMaxRewriteBytes` that are too large could cause the ingress controller to use significantly more memory than necessary.

Reviewer 2: Should ingress controller be capitalized here?
```suggestion
Setting this field not recommended because `headerBufferMaxRewriteBytes` values that are too small can break the Ingress Controller. Conversely, values for `headerBufferMaxRewriteBytes` that are too large could cause the Ingress Controller to use significantly more memory than necessary.
```

Author: No, I think that would confuse readers. MicroShift does not use the entire Operator so I don't want to imply that.

====

* `tuningOptions.headerBufferMaxRewriteBytes` specifies how much memory should be reserved, in bytes, from `headerBufferBytes` for HTTP header rewriting and appending for Ingress Controller connection sessions. The minimum value for `headerBufferMaxRewriteBytes` is `4096`. `headerBufferBytes` must be greater than `headerBufferMaxRewriteBytes` for incoming HTTP requests. If not set, the default value is `8192` bytes.
+
[IMPORTANT]
====

> Setting this field not recommended because `headerBufferMaxRewriteBytes` values that are too small can break the ingress controller. Conversely, values for `headerBufferMaxRewriteBytes` that are too large could cause the ingress controller to use significantly more memory than necessary.

Reviewer 2: Same question about capitalization for this admonition and the next one in this table cell
